extensions/amp-animation/0.1/web-animations.js

```diff
@@ -0,0 +1,363 @@
+      const object = /** {!Object<string, *>} */ (spec.keyframes);
+      keyframes = {};
+      for (const prop in object) {
+        const value = object[prop];
+        let preparedValue;
+        if (!isArray(value) || value.length == 1) {
+          // Missing "from" value. Measure and add.
+          const fromValue = this.measure_(target, prop);
+          preparedValue = [fromValue, clone(isArray(value) ? value[0] : value)];
+        } else {
+          preparedValue = clone(value);
+        }
+        keyframes[prop] = preparedValue;
+      }
+    } else if (isArray(spec.keyframes) && spec.keyframes.length > 0) {
+      // Keyframes -> property form.
+      const array = /** {!Array<!Object<string, *>>} */ (spec.keyframes);
+      keyframes = [];
+      // Fill-in partial key-frames.
+      // See https://github.com/web-animations/web-animations-js/issues/14
+      const addStartFrame = array.length == 1 || array[0].offset > 0;
+      const startFrame = addStartFrame ? {} : clone(array[0]);
+      keyframes.push(startFrame);
+      const start = addStartFrame ? 0 : 1;
+      for (let i = start; i < array.length; i++) {
+        const frame = array[i];
+        for (const prop in frame) {
+          if (!startFrame[prop]) {
+            // Missing "from" value. Measure and add to start frame.
+            startFrame[prop] = this.measure_(target, prop);
+          }
+        }
+        keyframes.push(clone(frame));
+      }
+    } else {
+      // No a known form of keyframes spec.
+      if (this.validate_) {
+        throw user().createError('keyframes not found', spec.keyframes);
+      }
+    }
+
+    this.requests_.push({target, keyframes, timing});
+  }
+
+  /** @override */
+  onUnknownAnimation(spec) {
+    if (this.validate_) {
+      throw user().createError('unknown animation type:' +
+          ' must have "animations" or "keyframes" field');
+    } else {
+      super.onUnknownAnimation(spec);
+    }
+  }
+
+  /**
+   * Runs the callback with the specified timing and restores old timing
+   * later.
+   * @param {!WebAnimationTimingDef} timing
+   * @param {function()} callback
+   * @private
+   */
+  withTiming_(timing, callback) {
+    const prevTiming = this.timing_;
+    this.timing_ = timing;
+    callback();
+    this.timing_ = prevTiming;
+  }
+
+  /**
+   * @param {string|!Element} targetSpec
+   * @return {!Element}
+   * @private
+   */
+  resolveTarget_(targetSpec) {
+    const target = user().assertElement(
+        typeof targetSpec == 'string' ?
+            this.context_.resolveTarget(targetSpec) :
+            targetSpec,
+        `Target not found: "${targetSpec}"`);
+    if (this.targets_.indexOf(target) == -1) {
+      this.targets_.push(target);
+    }
+    return target;
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {string} prop
+   * @private
+   */
+  measure_(target, prop) {
+    const index = this.targets_.indexOf(target);
+    if (!this.computedStyleCache_[index]) {
```

Comment:
I believe we need to invalidate this cache after animation run finishes otherwise a second animation run on specs that have fill-mode `forward` or `both` ( or if style changes with a user interaction between the first and second run ) could yield the wrong value